src/Generating/Trees.cpp

```diff
@@ -423,29 +478,26 @@ void GetLargeAppleTreeImage(int a_BlockX, int a_BlockY, int a_BlockZ, cNoise & a
 	// Trunk:
 	for (int i = 0; i < Height; i++)
 	{
-		a_LogBlocks.push_back(sSetBlock(a_BlockX, a_BlockY + i, a_BlockZ, E_BLOCK_LOG, E_META_LOG_APPLE));
+		a_LogBlocks.push_back(sSetBlock(a_BlockPos.addedY(i), E_BLOCK_LOG, E_META_LOG_APPLE));
 	}
 }
 
 
 
 
 
-void GetLargeAppleTreeBranch(int a_BlockX, int a_BlockY, int a_BlockZ, int a_BranchLength, Vector3d a_StartDirection, Vector3d a_Direction, int a_TreeHeight, cNoise & a_Noise, sSetBlockVector & a_LogBlocks)
+Vector3d GetTreeBranch(BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta, Vector3i a_BlockPos, int a_BranchLength, Vector3d a_StartDirection, Vector3d a_Direction, sSetBlockVector & a_LogBlocks)
 {
-	Vector3d CurrentPos = Vector3d(a_BlockX, a_BlockY, a_BlockZ);
+	Vector3d CurrentPos = Vector3d(a_BlockPos);
 	Vector3d Direction  = a_StartDirection;
 	for (int i = 0; i < a_BranchLength; i++)
 	{
 		CurrentPos += Direction;
-		if (CurrentPos.y >= a_TreeHeight)
-		{
-			return;
-		}
-		Direction -= a_Direction;
+		Direction += a_Direction;
 		Direction.Clamp(-1.0, 1.0);
-		a_LogBlocks.push_back(sSetBlock(FloorC(CurrentPos.x), FloorC(CurrentPos.y), FloorC(CurrentPos.z), E_BLOCK_LOG, GetLogMetaFromDirection(E_META_LOG_APPLE, Direction)));
+		a_LogBlocks.push_back(sSetBlock(FloorC(CurrentPos.x), FloorC(CurrentPos.y), FloorC(CurrentPos.z), a_BlockType, GetLogMetaFromDirection(a_BlockMeta, Direction)));
```

Comment:
Use `CurrentPos.Floor()` together with the shiny new `Vector3i`-based constructor :)